core/src/com/unciv/ui/trade/DiplomacyScreen.kt

```diff
@@ -258,6 +266,44 @@ class DiplomacyScreen(val viewingCiv:CivilizationInfo):CameraStageBaseScreen() {
         diplomacyTable.add(backButton)
         return diplomacyTable
     }
+    private fun getImprovementGiftTable(otherCiv: CivilizationInfo): Table{
+        val improvementGiftTable = getCityStateDiplomacyTableHeader(otherCiv)
+        improvementGiftTable.addSeparator()
+
+        val improvableTiles = otherCiv.getCapital().getImprovableTiles()
+        val tileImprovements = otherCiv.gameInfo.ruleSet.tileImprovements.filter { it.value.turnsToBuild != 0 }
+        var anyImprovements = false
+
+        for (improvableTile in improvableTiles){
+            for (tileImprovement in tileImprovements.values){
+                if (improvableTile.canBuildImprovement(tileImprovement, otherCiv)){
```

Comment:
Does this exclude 'cancel building improvement' and improvements that don't connect the resource on the tile?

Reply:
If a city state is building an improvement you can still buy the improvement. This does not exclude improvements that don't connect the resource on the tile but I will fix it.